### PR TITLE
add many nodes option

### PR DESCRIPTION
**Pull request by PrometheusPi**
This pr adds the option to distribute grid to multiple compute nodes (gpus in case of PIConGPU)

### CODE REVIEW

File: lib/python/picongpu/picmi/grid.py [hunk 1]
```diff
@@ -62,4 +72,29 @@ def get_as_pypicongpu(self):
             picongpu_boundary_condition_by_picmi_id[self.bc_ymin]
         g.boundary_condition_z = \
             picongpu_boundary_condition_by_picmi_id[self.bc_zmin]
+
+        # gpu distribution
+                # convert input to 3 integer list
```
Comment (s9105947) [inline lib/python/picongpu/picmi/grid.py]: indent

File: lib/python/picongpu/picmi/grid.py [hunk 2]
```diff
@@ -9,9 +9,19 @@
...
 import picmistandard
+from typeguard import typechecked
+import typing
 
 
+@typechecked
 class Cartesian3DGrid(picmistandard.PICMI_Cartesian3DGrid):
+    def __init__(self, n_gpus: typing.Optional[typing.List[int]] = None, **kw):
+        """overwriting PICMI init to extract gpu distribution for PIConGPU"""
+        self.n_gpus = n_gpus
```
Comment (s9105947) [inline lib/python/picongpu/picmi/grid.py]: behavior is not obvious, please add `:param n_gpus: description` in docstring

File: share/picongpu/pypicongpu/schema/grid.Grid3D.json [hunk 3]
```diff
@@ -41,6 +41,25 @@
...
+            "unevaluatedProperties": false,
+            "required": ["x", "y", "z"],
+            "properties": {
+                "x": {
+                    "$anchor": "gpu_cnt_component",
+                    "type": "integer",
+                    "minimum": 1
+                },
+                "y": {
+                    "$ref": "#cell_cnt_component"
```
Comment (s9105947) [inline share/picongpu/pypicongpu/schema/grid.Grid3D.json]: wrong anchor

File: share/picongpu/pypicongpu/schema/grid.Grid3D.json [hunk 4]
```diff
@@ -41,6 +41,25 @@
...
+                },
+                "z": {
+                    "$ref": "#cell_cnt_component"
```
Comment (s9105947) [inline share/picongpu/pypicongpu/schema/grid.Grid3D.json]: dito

File: share/picongpu/pypicongpu/template/etc/picongpu/N.cfg.mustache [hunk 5]
```diff
@@ -48,9 +48,11 @@ set -u
 
 TBG_wallTime="1:00:00"
 
-TBG_devices_x=1
-TBG_devices_y=1
-TBG_devices_z=1
+{{#grid.gpu_cnt}}
```
Comment (s9105947) [inline share/picongpu/pypicongpu/template/etc/picongpu/N.cfg.mustache:51]: maybe indent block?

File: test/python/picongpu/quick/pypicongpu/grid.py [hunk 6]
```diff
@@ -55,6 +56,24 @@ def test_types(self):
             g.boundary_condition_y = 1
         with self.assertRaises(TypeError):
             g.boundary_condition_z = {}
+        with self.assertRaises(TypeError):
+            g.n_gpus = [1, 1, 1] # list not accepted - tuple needed
+
+    def test_positiv(self):
```
Comment (s9105947) [inline test/python/picongpu/quick/pypicongpu/grid.py]: name: what is positive? -> `test_gpu_cnt_positive`
docstring: missing

File: test/python/picongpu/quick/pypicongpu/grid.py [hunk 7]
```diff
@@ -55,6 +56,24 @@ def test_types(self):
...
+        g = self.g
+        with self.assertRaisesRegex(Exception, ".*greater than 0.*"):
```
Comment (s9105947) [inline test/python/picongpu/quick/pypicongpu/grid.py]: include axis in regex (!! important, will lead to more issues)

File: test/python/picongpu/quick/picmi/grid.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            number_of_cells=[192, 2048, 12],
+            lower_bound=[0, 0, 0],
+            upper_bound=[3.40992e-5, 9.07264e-5, 2.1312e-6],
+            lower_boundary_conditions=['open', 'open', 'periodic'],
+            upper_boundary_conditions=['open', 'open', 'periodic'])
+
+    def test_basic(self):
+        """simple translation"""
+        grid = self.grid
+        grid.get_as_pypicongpu()
```
Comment (s9105947) [inline test/python/picongpu/quick/picmi/grid.py]: add get_rendering_context() for full check suite

File: test/python/picongpu/quick/picmi/grid.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    def test_typo_ngpus(self):
+        """test common typo ngpus instead of n_gpus"""
+        with self.assertRaisesRegex(TypeError, "Unexpected keyword argument: \"\[\'ngpus\'\]\""):
```
Comment (s9105947) [inline test/python/picongpu/quick/picmi/grid.py]: `.*unexpected.*ngpus.*` enough? -> for readability

File: test/python/picongpu/quick/picmi/grid.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            grid = picmi.Cartesian3DGrid(
+                number_of_cells=[192, 2048, 12],lower_bound=[0, 0, 0],
```
Comment (s9105947) [inline test/python/picongpu/quick/picmi/grid.py]: formatting

**Review by s9105947 — CHANGES_REQUESTED** [review]
as dicussed offline

**Comment by s9105947** [comment]
add doc in `docs/source/usage/picmi/intro.rst` on PIConGPU-exclusive PICMI extension `n_gpus`

File: share/picongpu/pypicongpu/template/etc/picongpu/N.cfg.mustache (hunk 5 above)
Comment (PrometheusPi) [reply share/picongpu/pypicongpu/template/etc/picongpu/N.cfg.mustache:51]: will do and indent you blocks as well

File: test/python/picongpu/quick/pypicongpu/grid.py (hunk 6 above)
Comment (PrometheusPi) [reply test/python/picongpu/quick/pypicongpu/grid.py]: will add doc strings to all other tests as well

**Comment by PrometheusPi** [comment]
@s9105947 Done - thanks for the review

**Review by s9105947 — COMMENTED** [review]
thx

**Review by s9105947 — APPROVED** [review]

**Pull request merged**